v1_1/system.ts

```diff
@@ -0,0 +1,167 @@
+ * Location is used by a location owner to create the two QRCodes.
+ */
+export class Location {
+    public data: ILocationData;
+    public notificationKey = randombytes_buf(32);
+
+    constructor(public healthAuthorityPubKey: Uint8Array,
+                public name: string,
+                public location: string,
+                public room: string,
+                public locationType: number) {
+        this.data = CryptoV1_1.genCode(this.infoStr());
+    }
+
+    static infoStr(name: string, location: string, room: string, locationType: number): string {
+        return [name, location, room, locationType.toString()].join("::");
+    }
+
+    infoStr(): string {
+        return Location.infoStr(this.name, this.location, this.room, this.locationType);
+    }
+
+    /**
+     * Returns the base64 encoded protobuf-message for the location owner.
+     *
+     * @param baseURL - anything - is ignored and removed afterwards.
+     */
+    getQRtrace(baseURL: string): string {
+        const qrct = new QRCodeTrace({
+            version: 2,
+            trSk: this.data.tr.sk,
+            trR2: this.data.tr.r2,
+            pTrR1: this.data.pTr.r1,
+            pTrR2: this.data.pTr.r2,
+            info: this.infoStr(),
+        });
+        const traceBuf = QRCodeTrace.encode(qrct).finish();
+        const traceEnc = crypto_box_seal(traceBuf, this.healthAuthorityPubKey);
+        return `${baseURL}#${to_base64(traceEnc)}`;
+    }
+
+    /**
+     * Returns the base64 encoded protobuf-message necessary for visitors to register.
+     *
+     * @param baseURL - anything - is ignored and removed afterwards.
+     */
+    getQRentry(baseURL: string): string {
+        const qrCodeContent = QRCodeContent.create({
+            version: 2,
+            publicKey: this.data.ent,
+            pEnt: this.data.pEnt,
+            name: this.name,
+            location: this.location,
+            room: this.room,
+            venueType: this.locationType,
+            notificationKey: this.notificationKey
+        });
+
+        const qrCodeBuf = QRCodeContent.encode(qrCodeContent).finish();
+        return `${baseURL}#${to_base64(qrCodeBuf)}`;
+    }
+}
+
+/**
+ * The user has zero or more instances of Visit in his phone.
+ */
+export class Visit {
+    private visit: IVisit;
+
+    constructor(
+        qrCodeEntry: string,
+        public entry: number,
+        public departure: number,
+        diary?: boolean
+    ) {
+        const qrBase64 = qrCodeEntry.replace(/^.*#/, '');
+        const content = QRCodeContent.decode(from_base64(qrBase64));
+        if (content === undefined) {
+            throw new Error("didn't find 'content'");
+        }
+        if (content.version !== 2) {
+            throw new Error("Wrong version")
+        }
+        const infoBuf = Location.infoStr(content.name, content.location, content.room,
+            content.venueType);
+        this.visit = CryptoV1_1.scan(content.publicKey, content.pEnt, infoBuf, diary);
```

Comment:
Note how `entry` and `departure` are never stored in scan

Reply:
Also, probably want to pass in `notification_key`. As above, I would be inclined to either pass `aux` as something you format here (it would include `notification_key`), or as in v1 (better for seeing parallel): pass `content`/`info` raw, and do processing inside `scan`

Reply:
fixed